### PR TITLE
add kubeconfig types

### PR DESCRIPTION
**Pull request by deads2k**
Adds the ability to read a kubeconfig file (issue #1755).

Addresses #2644 to allow binding to different names and provides suggested names with a user specified prefix.

### CODE REVIEW

File: pkg/client/clientcmd/client_builder.go [hunk 1]
```diff
@@ -17,186 +17,123 @@ limitations under the License.
...
+// NewInteractiveKubeConfig creates a DirectKubeConfig using the passed context name and a reader in case auth information is not provided via files or flags
+func NewInteractiveKubeConfig(config Config, contextName string, overrides *ConfigOverrides, fallbackReader io.Reader) KubeConfig {
+	return DirectKubeConfig{config, contextName, overrides, fallbackReader}
 }
 
-// Config implements Builder
-func (builder *builder) Config() (*client.Config, error) {
-	if builder.config != nil {
-		return builder.config, nil
+// Client implements KubeConfig
```
Comment (jlowdermilk) [inline pkg/client/clientcmd/client_builder.go]: Forget to delete this?

File: pkg/client/clientcmd/loader_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"gopkg.in/v2/yaml"
+)
+
+var (
+	testConfigAlfa = Config{
```
Comment (jlowdermilk) [inline pkg/client/clientcmd/loader_test.go]: Totally optional nit: s/alfa/alpha

**Comment by jlowdermilk** [comment]
LGTM, this fits the design proposed in #1755 nicely. Just nits above.

Might be helpful to add a .md doc that includes an explanation of load order. Maybe cut/paste the text of the comment on [getKubeConfig](https://github.com/deads2k/kubernetes/blob/deads-add-clicfg-types/pkg/kubectl/cmd/cmd.go#L132)

File: pkg/client/clientcmd/client_builder.go (hunk 1 above)
Comment (deads2k) [inline pkg/client/clientcmd/client_builder.go]: > Forget to delete this?

Yes.  Tidied up.

File: pkg/client/clientcmd/loader_test.go (hunk 2 above)
Comment (deads2k) [inline pkg/client/clientcmd/loader_test.go]: > Totally optional nit: s/alfa/alpha

Nato phonetics (alfa, bravo) as opposed to Greek letters (alpha, beta).  Juliett is spelled weird too, but I haven't gotten that far in recent tests: http://en.wikipedia.org/wiki/NATO_phonetic_alphabet.

**Comment by deads2k** [comment]
> Might be helpful to add a .md doc that includes an explanation of load order.

@jlowdermilk Just a doc under kubernetes/docs?

**Comment by deads2k** [comment]
@jlowdermilk Added https://github.com/deads2k/kubernetes/blob/deads-add-clicfg-types/docs/kubeconfig-file.md to describe the loading order.

**Comment by smarterclayton** [comment]
Some final comments (one substantial concern about validation, one gap with flag names), otherwise is in good shape to me.

**Comment by deads2k** [comment]
@smarterclayton Got the renames.  I put the error type export in a separate commit so you can confirm that's how you want it.

**Comment by deads2k** [comment]
@smarterclayton tweaked, rebased, squashed.

**Comment by smarterclayton** [comment]
LGTM

**Comment by brendandburns** [comment]
I _think_ that this broke e2e because we didn't make changes to cluster turn up to write the appropriate config file.

**Comment by brendandburns** [comment]
I would lean towards rolling this back, and then have a re-revert with changes to kube-up function cluster/*/ 

Sorry!

**Comment by deads2k** [comment]
@brendanburns All command line flags were kept exactly the same and the file isn't supposed to be required, however vagrant e2e has always been hit or miss so I got the same results as I got on master and considered it good.

Would you like me to do the revert?  Also, do you have a gist with the failure?

**Comment by zmerlynn** [comment]
The e2e failure looks about like this:

```
[0;32mKubernetes cluster is running.  The master is running at:

[0;33m  https://130.211.187.55

[0;32mThe user name and password to use is located in ~/.kubernetes_auth.[0m

... calling validate-cluster
Project: kubernetes-jenkins
Project: kubernetes-jenkins
Running: /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../cluster/../cluster/gce/../../cluster/../cluster/gce/../../platforms/linux/amd64/kubectl get minions -o template -t {{range.items}}{{.id}}
{{end}}
F0107 20:34:08.663798   22537 get.go:75] Get https://130.211.187.55/api/v1beta1/minions?namespace=default: x509: certificate signed by unknown authority
2015/01/07 20:34:08 Error running up: exit status 255
2015/01/07 20:34:08 Error starting e2e cluster. Aborting.
exit status 1
```

**Comment by deads2k** [comment]
Ok, well that's pretty bad.  I see the problem.  .kubernetes_auth contains information for the server to identify the user and information for the client to identify the server.  I lost the latter half during a refactor.

I'll put together a change to restore that and show that succeeding.  I'm sorry about this.

**Comment by deads2k** [comment]
@zmerlynn Forgive my ignorance, but is there an easy way for me to run e2e using the gke provider from outside google (getting started lists gce, but not gke)?

While there was a bug in respecting "Insecure" from a .kubernetes_auth file, that didn't affect the vagrant environment either way and I noticed that "--auth_path" is handled differently for gke depending on whether it is kubecfg of kubectl.  Basically, it means that making sure auth for e2e runs properly vagrant (It did before too) doesn't mean that it will work on gke.

Alternatively, if the ~/.kubernetes_auth from your failing test had "Insecure":true, then we're golden.

**Pull request merged**